### DISCUSSION TITLE
non-eval and/or lazy versions of dataclasses/namedtuple methods

### DISCUSSION THREAD

**Discussion by RonnyPfannschmidt** [faster-cpython/ideas#195] [Ideas]
as utilities like data-classes/namedtuples use eval to compose their helpers,
larger libraries pay a regular repeated cost of parse, compile, eval for their components

in https://github.com/dabeaz/dataklasses @dabeaz demonstrated a rather refreshing variant of the possible speed for a number of cases

in a ipython notebook i made some pretty rough, but seemingly accurate enough comparisons

```python
import attr
from dataclasses import dataclass
from dataklasses import dataklass

```

```python
%%timeit
@attr.s(auto_attribs=True)
class funa:
    name: str
    age: int|None = None
        

        
    
```

    28.7 ms ± 226 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

```python
%%timeit
@dataclass
class funa:
    name: str
    age: int|None = None
        
```

    253 µs ± 5.18 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

```python
%%timeit
@dataklass
class funa:
    name: str
    age: int|None = None
```

    11 µs ± 132 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

```python
%%timeit
class funp:
    name: str
    age: int|None
    def __init__(self, name: str, age: int|None = None):
        self.name = name
        self.age = age
        
    def __eq__(self, other):
        if not isinstance(other, funp):
            return NotImplemented
        return self.name == other.name and self.age == other.age
    def __repr__(self):
        return f"<fun name={self.name} age={self.age})"
    
    
```

    5.43 µs ± 268 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

**Comment by gvanrossum** [comment]
The whole internet is talking about it. :-)

I’ve already pointed @ericvsmith at it, if it’s viable I’m sure it will happen.

**Reply by RonnyPfannschmidt** [reply]
a non-explored variant is to replace the python/codegen based implementations with c level ones
which wouldn't incur any of the cost at any of the times - however i do suspect some resistance

the devils advocate in me however wants to throw in that nobody seems to be "missing" a python level implementation of list or int either
